jax/_src/nn/initializers.py

```diff
@@ -126,7 +126,7 @@ def init(key: KeyArray,
            shape: core.Shape,
            dtype: DTypeLikeInexact = dtype) -> Array:
     dtype = dtypes.canonicalize_dtype(dtype)
-    return random.uniform(key, shape, dtype) * scale
+    return random.uniform(key, shape, dtype) * lax.convert_element_type(scale, dtype)
```

Comment:
I don't think this is correct: if a user has `jax_numpy_dtype_promotion=strict` set and the dtypes do not match, this should raise an error.

Reply:
Same with all the new `convert_element_type` calls below.

Reply:
OK IIUC you are saying that the current behavior is intended.

In the example in #18267
```
# os.environ["JAX_DEFAULT_DTYPE_BITS"] = "32"
os.environ["XLA_PYTHON_CLIENT_MEM_FRACTION"] = "0.7"
os.environ["JAX_ENABLE_X64"] = "1"
import jax
import jax.numpy as jnp
import numpy as np

_default_kernel_init = jax.nn.initializers.orthogonal(np.sqrt(2), dtype=jnp.float32)
param = _default_kernel_init(jax.random.PRNGKey(0), (5,5), jnp.float32)
print(param.dtype) # float64
```
Then you were saying that this is the intended behavior right.

More generally, for any initializer that takes a scale parameter, then if the scale uses a numpy array of dtype float64, then we expect the resulting array dtype to be float64 and even when user requested float32 as the parameter dtype. Is that correct?

Reply:
Sorry - lost track of this.

No, the current behavior is not necessarily intended, but it is intended that passing mixed-precision inputs under strict dtype promotion should result in an error, and we should not sidestep that promotion error by hard-casting values.

Reply:
I am trying to figure out what the right thing should be since the use case I have in #18267 should indeed will trigger the error under strict dtype promotion. Are you saying that we should also warn/error the user about the potential upcasting in that case?

Reply:
The main problem with this PR as currently written is that when a user does not pass an explicit dtype, it still casts `scale` to the default `dtype`. We could change the default dtype to `None` and only do this casting when there actually is a user-specified dtype. What do you think?

Reply:
Sounds good to me, but this probably requires a big change to all initializers though (essentially replacing every dtype in the outer function with None).

I implemented this behavior for uniform in f6a79b4, if you think this is what we want I will update the other initializers. Don't worry about the tests for now, I will do some clean up later for review.